File: src/bin/day06.rs
#![feature(iterator_fold_self)]

extern crate nom;
use std::collections::HashSet;
use std::fs;

fn main() {
    let contents = fs::read_to_string("inputs/day06.txt").expect("Something went wrong reading the file");

    let sum_anyone: usize = contents
        .split("\n\n")
        .map(|group| {
            group
                .lines()
                .flat_map(|ln| ln.chars())
                .collect::<HashSet<_>>()
                .len()
        })
        .sum();
    println!("Part 1: {}", sum_anyone);

    let sum_everyone: usize = contents
        .split("\n\n")
        .map(|group| {
            group
                .lines()
                .map(|ln| ln.chars().collect::<HashSet<_>>())
                .fold_first(|acc, ln| acc.intersection(&ln).cloned().collect())
                .unwrap()
                .len()
        })
        .sum();
    println!("Part 2: {}", sum_everyone);
}

File: julia/day06.jl
input = open("../inputs/day06.txt") do f
    read(f, String)
end

groups = [
    [
        Set([c for c in ln])
        for ln in split(group, "\n")
    ]
    for group in split(input, "\n\n")
]

part1 = sum([length(union(group...)) for group in groups])
println("Part 1: ", part1)

part2 = sum([length(intersect(group...)) for group in groups])
println("Part 2: ", part2)
